[SYCL][Fusion] Adapt internalization to opaque pointers.

jopperm opened the pull request:
Fixes `sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp` which was temporarily deactivated in #9828.

Switching to opaque pointers meant that accesses to the data structure in the testcase were no longer represented as one "deep" GEP, but rather a series of GEPs into the different nesting levels of the structure. Rather than bailing out, we now keep track of GEPs that index into an aggregate object, and exclude their users from the modulo-`LocalSize` rewriting (the pointer types still have to be changed to the desired target address space, though).

In addition, perform `-simplifycfg` before internalization in order to get rid of unreachable code which, in this particular case, contained Phi nodes on the candidate accessor.

sommerlukas commented on sycl-fusion/passes/internalization/Internalization.cpp:
```diff
@@ -232,27 +241,20 @@ Error SYCLInternalizerImpl::canPromoteCall(CallBase *C, const Value *Val,
   // Recurse to check all users of the GEP.
-  return canPromoteValue(GEPI, LocalSize);
+  return canPromoteValue(GEPI, LocalSize,
+                         InAggregate || GEPI->getNumIndices() >= 2);
```
Can we document why `2` is the magic number here?

sommerlukas reviewed (commented):
Do we need to add a LIT test checking that we do not promote a pointer into an aggregate?

jopperm commented:
CI failure looks unrelated. I'm still missing a testcase for aborting propagation into function calls, though.

jopperm commented:
I've added a lit test for stopping the promotion of calls in `InAggregate` mode, and changed the corresponding error message to an NYI (rather than "it's unsafe") -- I think it could be supported, but we'd need to change the implementation to propagate the `InAggregate` flag on a per-argument-basis, which I'd argue is out of scope for this PR.

jopperm commented:
Failure on CUDA is caused by a test added by this PR; investigating.
